sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/objects.scala

```diff
@@ -96,3 +99,40 @@ object CombineTypedFilters extends Rule[LogicalPlan] {
     }
   }
 }
+
+/**
+ * Removes MapObjects when the following conditions are satisfied
+ *   1. Mapobject(e) where e is lambdavariable
+ *   2. the function will convert an expression MapObjects(e) to AssertNotNull(e)
+ *   3. the inputData is of primitive type array and its element is not nullable.
+ *   4. the outputData is of primitive type array and its element does not have enull
+ *   5. no custom collection class specified
+ * representation of data item.  For example back to back map operations.
+ */
+object EliminateMapObjects extends Rule[LogicalPlan] {
+  private def convertDataTypeToArrayClass(dt: DataType): Class[_] = dt match {
+    case IntegerType => classOf[Array[Int]]
+    case LongType => classOf[Array[Long]]
+    case DoubleType => classOf[Array[Double]]
+    case FloatType => classOf[Array[Float]]
+    case ShortType => classOf[Array[Short]]
+    case ByteType => classOf[Array[Byte]]
+    case BooleanType => classOf[Array[Boolean]]
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case _ @ DeserializeToObject(_ @ Invoke(
+          MapObjects(_, _, inputType, args, inputData, customCollectionCls, _),
```

Comment:
I think we can just match `MapObjects`, as it's quite self-contained, we don't need other information to decide if we can remove `MapObjects`.